web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,400 @@
+        if (interval.startTime >= lastEndTime) {
+          const intervalInData = this.data.find(({ pushID }) => pushID === interval.pushID);
+          if (intervalInData) {
+            intervalInData.row = rowIndex;
+          }
+          lastEndTime = interval.endTime;
+        } else {
+          overlappingIntervals.push(interval);
+        }
+      }
+      rowIndex++;
+      lastEndTime = 0;
+      data = overlappingIntervals;
+    }
+    this.numRows = rowIndex;
+  }
+
+  /**
+   * Converts the date (in milliseconds) to a datetime string. The returned
+   * date will have a format of YYYY-MM-DD hh-mm, with single digits padded
+   * with a leading zero.
+   * @param d Holds one interval's data on the timeline.
+   */
+  private formatDate = (d: any) => {
+    const date = new Date(d);
+    const dateTimeFormat = new Intl.DateTimeFormat('en', {
+      year: 'numeric',
+      month: '2-digit',
+      day: '2-digit',
+      hour: '2-digit',
+      minute: '2-digit'
+    });
+    const [{ value: month }, , { value: day }, , { value: year }, , { value: hour }, , { value: minute }] = dateTimeFormat['formatToParts'](date);
+    return `${year}-${month}-${day} ${hour}:${minute}`;
+  };
+
+  /**
+   * Composes content of the tooltip that will appear on hover.
+   * @param d Holds one interval's data on the timeline.
+   */
+  private getTooltipContent = (d: Item) => {
+    const duration = (d.endTime - d.startTime);
+    const options = ({
+      round: true
+    } as HumanizeDurationOptions)
+    const output = TimelineComponent.HUMANIZER.humanize(duration, options);
+
+    // HTML representation of all required data.
+    return `<b>Push ID: ${d.pushID.slice(d.pushID.indexOf('@') + 1)}</b>
+      <br/>
+      <b>Final State: ${d.state}</b>
+      <br/>
+      <b>Start Time: ${this.formatDate(d.startTime)}</b>
+      <br/>
+      <b> End Time: ${this.formatDate(d.endTime)}</b>
+      <br/>
+      <b>Duration: ${output}</b>
+      `
+  };
+
+  /**
+   * Create tooltip to display each interval's content.
+   * @param el Encasing element that holds the tooltip
+   */
+  private createTooltip = (el: any) => { // TODO: tachyons?
+    el.style('position', 'absolute')
+      .style('pointer-events', 'none')
+      .style('top', 0)
+      .style('opacity', 0)
+      .style('background', 'white')
+      .style('border-radius', '5px')
+      .style('box-shadow', '0 0 10px rgba(0,0,0,.25)')
+      .style('padding', '10px')
+      .style('line-height', '1.3')
+      .style('font', '11px sans-serif');
+  };
+
+  /** 
+   * Defines the zoom behavior, limiting the scale with which we can zoom in and out
+   * and restricting zoom to the x-axis only. Upon zoom, the x-axis will rescale,
+   * as will the timeline intervals.
+   */
+  private zoom = d3.zoom()
+    .scaleExtent([0.75, 1000]) // Limit zoom out.
+    .translateExtent([[-100000, 0], [100000, 0]]) // Avoid scrolling too far.
```

Comment:
Should we be setting these values arbitrarily or based on the data? What if there's one push def that has lots of data spread out across years vs a timeline of just one push?

Reply:
Gotcha, scaled it down to the 5 second mark on the axis, which should be enough.